### ISSUE TITLE
How can you decrypt the signature?

### ISSUE THREAD

**Issue by KhanhNB** [flagbug/YoutubeExtractor#162]
I know that with normal videos, Youtube will hardcode the signature in links in "url_encoded_fmt_stream_map", but with some videos like VEVO videos, they will only provide one signature so the links in url_encoded_fmt_stream_map will not work. How could know 'how to decrypt  these signatures' ? Can you share it? Thank You!

**Comment by flagbug** [comment]
Call `DecryptDownloadUrl` https://github.com/flagbug/YoutubeExtractor/blob/1c6d99c916f4479af628e69a4390690c6c412d1b/YoutubeExtractor/YoutubeExtractor/DownloadUrlResolver.cs#L28

**Comment by KhanhNB** [comment]
No, I mean how did you know "how Youtube encrypted it"? And how to find the way to decrypt it?

**Comment by flagbug** [comment]
No idea, when I initially created YoutubeExtractor I just gathered code from various sources and looked into the https://github.com/rg3/youtube-dl source code. I don't know how @Kimmax is doing these things, maybe he can tell you more

**Comment by KhanhNB** [comment]
Thank you very much :+1:

**Comment by Kimmax** [comment]
Hey there!
I do this by taking the existing code and in case of a change (like the one which was fixed in 2fb0b8d2abdce999df7e8b439597fd6e24b7f26c) try to find out at what point the code fails. And then do some reverse engineering to get the code working again. Let's take as an example the latest code change.

The most important code for this is the following _(from Decipherer.cs, Method DecipherWithVersion)_
```C#
string jsUrl = string.Format("http://s.ytimg.com/yts/jsbin/player-{0}.js", cipherVersion);
string js = HttpHelper.DownloadString(jsUrl); // ### BREAKPOINT ###

//Find "C" in this: var A = B.sig||C (B.s)
string functNamePattern = @"\.sig\s*\|\|([a-zA-Z0-9\$]+)\("; //Regex Formed To Find Word or DollarSign

var funcName = Regex.Match(js, functNamePattern).Groups[1].Value;
            
if (funcName.Contains("$")) 
{
    funcName = "\\" + funcName; //Due To Dollar Sign Introduction, Need To Escape
}

string funcPattern = @"var " + @funcName + @"=function\(\w+\)\{.*?\};"; //Escape funcName string
var funcBody = Regex.Match(js, funcPattern).Value; //Entire sig function
var lines = funcBody.Split(';'); //Each line in sig function
```

Start by placing your breakpoint at the line where the player js is being stored after being downloaded, in this case

```C#
 string js = HttpHelper.DownloadString(jsUrl);
```

And check if this variable get's filled with data _(Tons of, actually)_. If this is **not** the case you have found the first problem: The player file is located in a different location. If this is the case most likely the Regex in the 
```C#
GetHtml5PlayerVersion()
```
 method has to be updated.
Got the js content in your variable? Good! Copy it over to some Regex capable text editor. _(Sublime does a good job at this)_.
Now continue step for step in VS.
  
The next thing to check is
```C#
//Find "C" in this: var A = B.sig||C (B.s)
string functNamePattern = @"\.sig\s*\|\|([a-zA-Z0-9\$]+)\(";
```
Notice the comment _Find "C" in this: var A = B.sig||C (B.s)_.
None of this characters are actually being used, except for _.sig_. All other are placeholders and are there to show the general form of what we search.
What we search for is this:
A variable (Keyword _var_ in js) with the name _A_ which checks if object _B.sig_ OR the return value of method _C_ which gets parameter _B.s_ passed is true (Which means the signature needs to be decrypted).
Now this one was hard, but don't care as much. What we need to know is the actual name of the method **C**, which does the decryption if run in the browser, so we can use it's code.
To do this we have a regex witch searches exactly for this:
```Regex
\.sig\s*\|\|([a-zA-Z0-9\$]+)\(
```
which does this (Copied from regex101.com)
```
\.sig\s*\|\|([a-zA-Z0-9\$]+)\(/
\. matches the character . literally
sig matches the characters sig literally (case sensitive)
\s* match any white space character [\r\n\t\f ]
Quantifier: * Between zero and unlimited times, as many times as possible, giving back as needed [greedy]
\| matches the character | literally
\| matches the character | literally
1st Capturing group ([a-zA-Z0-9\$]+)
[a-zA-Z0-9\$]+ match a single character present in the list below
Quantifier: + Between one and unlimited times, as many times as possible, giving back as needed [greedy]
a-z a single character in the range between a and z (case sensitive)
A-Z a single character in the range between A and Z (case sensitive)
0-9 a single character in the range between 0 and 9
\$ matches the character $ literally
\( matches the character ( literally
```
I let out the dollar sign because this is deprecated and not used anymore.
If you use this regex on the js file (case-sensitive) you should find exactly one match _(hopefully)_ in this case it's
```
.sig||gr(
```
If you add some characters before and at the end of the match you can see the form again:
```
var f=e.sig||gr(e.s)
```
But this time we have the real name of the function which holds the code for decrypting the signatures **gr**.
Since we have the function name now we should be able to pull the code of the method, right? Let's start of with the regex to find that:
```
var gr=function\(\w+\)\{.*?\};
```
again, the explanation:
```
var gr=function matches the characters var gr=function literally (case sensitive)
\( matches the character ( literally
\w+ match any word character [a-zA-Z0-9_]
Quantifier: + Between one and unlimited times, as many times as possible, giving back as needed [greedy]
\) matches the character ) literally
\{ matches the character { literally
.*? matches any character (except newline)
Quantifier: *? Between zero and unlimited times, as few times as possible, expanding as needed [lazy]
\} matches the character } literally
; matches the character ; literally
```
This is quite simple: First find the function declaration _var gr=function_ add the parameter (could be everything we don't care, but only word characters are allowed -> _(\w+)_ now the beginning of the code block _{_, now everything in it (the function's code!) but only up to the end of the code block _}_ and end of declaration _;_.
Should work right?
No this time. Because that's what changed. If you do step by step debugging you'll find out that after the
```C#
 var funcBody = Regex.Match(js, funcPattern).Value; //Entire sig function
```
call the variable stay empty, because the regex did not match anything in the js file (you can try yourself in your editor). That's bad. What we need to do now is some searching to find the code block in the js again.
You may do in a other way, but I started this way:
Simplify the regex until you get a reasonable amount of matches to check manually. (Don't try to optimize when you are down to 10 matches if this would take you longer to do than checking manually).
I started by dropping any unneeded characters.
First I dropped _var_. No match.
Dropped _;_. No match.
Dropped everything till I got to _gr=_. If searching case-sensitive you come down to 3 matches -> check them.
1)
```javascript
c.cgr=!0;
```
2) 
```javascript
h.gr=function(){W8(this,this.ha,this.Y)};
```
3) 
```javascript
gr=function(a){a=a.split("");
fr.Ww(a,13);fr.W9(a,69);fr.Gz(a,3);fr.Ww(a,2);fr.W9(a,79);fr.Gz(a,3);fr.Ww(a,36);return a.join("")}
``` 
No look at the code itself.
1) Not a function declaration, can't be it.
2) Is a function, but takes no parameter's (Remember _C_ takes _B.s_ as parameter), can't be it
3) Is a function, takes parameter, does some actual work and returns something. **Bingo**.

Now we need a new regex to match the new location and format of the function, look before and after the function:
```javascript
substr(2))},gr=function(a){a=a.split("");
fr.Ww(a,13);fr.W9(a,69);fr.Gz(a,3);fr.Ww(a,2);fr.W9(a,79);fr.Gz(a,3);fr.Ww(a,36);return a.join("")},hr=function(a
```
What the content does is not our business, but the formatting before and after is:
It looks like the declaration of the function is now in a list or similar, so the _var_ got dropped, also the semicolon at the end got replaced by an comma. Now adapt these changes to our regex we end up with something like this:
```
gr=function\(\w+\)\{.*?\},
```
also it look like there were some newlines added so it's a good idea to tell regex to match over the whole content, not only per line (Not necessarily necessary, but it could safe us from some trouble in the future)

Bring the regex back into our decryption method:
```C#
string funcPattern = @funcName + @"=function\(\w+\)\{.*?\},"; //Escape funcName string
``` 
and voilà _funcBody_ should have a value again.
Case closed.

From here it's taking what's in the function body, understanding what is done with the signature (No real encryption, just some character swapping and string reversing) and translating it to C#, so not only our webbrowser can do this, but also our own little library. But this would be even more too much than this comment here is, but googling "youtube signature decryption" should lead you way.

Have a great day!

**Comment by KhanhNB** [comment]
@Kimmax thank you very much :+1:
